### ISSUE TITLE
Ensure "Pod Portability" while using secrets-store-csi-driver

### ISSUE THREAD

**Issue by saad-ali** [deislabs/secrets-store-csi-driver#42]
tldr: One suggestion for improvement: ensure "pod portability" across clusters while using this driver.

**Problem**

This driver asks users to specify the secret provider inline in pod definition. For example:

```YAML
volumes:
  - name: secrets-store-inline
    csi:
      driver: secrets-store.csi.k8s.com
      readOnly: true
      volumeAttributes:
        providerName: "azure" <-- Avoid this
```

The problem with that is, if you use this pod definition and decide to move your pod to a non-azure cluster, you'll have to update your pod.

**Background**

In general, workload (pod) portability is something we try very hard to maintain in Kubernetes. With persistent volumes, for example, we discourage specifying a specific volume (e.g. AzureDisk or GCE PD) inline in a pod definition, because it makes the pod definition non-portable: i.e. if you move a pod directly referencing a GCE PD to a non-Google cluster, the pod won't work. So for persistent volumes, we encourage the use of PVC/PV so that your pod/PVC definition remains portable across clusters.

In this case, since this driver is ephemeral, PV/PVC wouldn't be appropriate.

**Suggestion**

One way to address this would be to create a layer of abstraction similar to a StorageClass for secret providers. Maybe call it `SecretProviderClass`? Similar to StorageClass, this new non-namespaced `SecretProviderClass` object would specify 1) the secret provider to use and 
2) the (opaque) parameters, if any, to pass to it during provisioning. For example:

```YAML
kind: SecretProviderClass
metadata:
  name: my-secret-provider
provisioner: azure
parameters:
  someStoreSpecificParameter: foo
```

On the pod side, instead of referencing specific `providerName`, the user would specify the name of a `secretProviderClass`:

```YAML
volumes:
  - name: secrets-store-inline
    csi:
      driver: secrets-store.csi.k8s.com
      readOnly: true
      volumeAttributes:
        secretProviderClass: "my-secret-provider"
        ...
```

The pod `volumeAttributes` should only contain paramters that are generic and are understood by any secret provider (values here can be unique per pod). This assumes that you can define a generic way to identify a secret (and any other pod specific fields that you need).

The `SecretProviderClass.parameters` should can contain provider specific parameters (values here are can not be specific per provider).

This way, the user could move the pod object from one cluster to another, and not worry about who the secret provider is. As long as there is a `SecretProviderClass` named `my-secret-provider`, things will just work.

**Comment by ritazh** [comment]
Thanks for the feedback @saad-ali. To ensure "pod portability", do you recommend creating a custom resource of type `SecretProviderClass` over creating PV/PVCs? Users currently can reference a PVC in the pod definition https://github.com/deislabs/secrets-store-csi-driver/blob/master/deploy/example/nginx-pod-secrets-store.yaml#L16 and define csi driver properties in the PV definition: https://github.com/deislabs/secrets-store-csi-driver/blob/master/deploy/example/pv-secrets-store-csi.yaml

**Comment by saad-ali** [comment]
PV/PVC definitely achieves pod portability. But it's a clunky interface. User has to create a PV object (which is non-namespaced) and many clusters restrict access to non-namespaced to admins only.

**Comment by ritazh** [comment]
Thanks! will work on this soon!
